modules/renter/proto/contractset.go

```diff
@@ -22,11 +22,34 @@ type ContractSet struct {
 	mu        sync.Mutex
 }
 
-// Len returns the number of contracts in the set.
-func (cs *ContractSet) Len() int {
+// Acquire looks up the contract with the specified FileContractID and locks
+// it before returning it. If the contract is not present in the set, Acquire
+// returns false and a zero-valued RenterContract.
+func (cs *ContractSet) Acquire(id types.FileContractID) (modules.RenterContract, bool) {
 	cs.mu.Lock()
-	defer cs.mu.Unlock()
-	return len(cs.contracts)
+	safeContract, ok := cs.contracts[id]
+	cs.mu.Unlock()
+	if !ok {
+		return modules.RenterContract{}, false
+	}
+	safeContract.mu.Lock()
```

Comment:
This seems like a deadlock waiting to happen.  Is there some way to ensure that this gets unlocked?

Reply:
The reason we can't currently merge this PR is that the concurrency paradigms around the contract set are not sufficient. We can either choose between race conditions or deadlocks, or really slow performance.

The future work on this branch is moving the contracts fully into the `proto` package, which will allow us to avoid the deadlocks and race conditions without hurting performance.